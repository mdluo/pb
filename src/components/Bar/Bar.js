import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Progress } from 'antd';
import classnames from 'classnames';

import styles from './Bar.module.less';

class Bar extends PureComponent {
  state = {
    percent: 0,
    value: 0,
  }

  static getDerivedStateFromProps(props) {
    if (props.limit <= 0) {
      throw new Error('The `limit` prop should be greater than 0.');
    }
    let percent = 0;
    const value = Math.max(props.value, 0);
    if (props.value > 0) {
      percent = Math.floor((props.value / props.limit) * 100);
    }
    return { percent, value };
  }

  render() {
    const { active, height } = this.props;
    const { percent, value } = this.state;
    return (
      <div className={classnames(styles.bar, { active }, { full: percent >= 100 })}>
        <Progress
          strokeWidth={height}
          showInfo={false}
          percent={percent}
          status={percent > 100 ? 'exception' : null}
        />
        <div
          className={styles.reading}
          style={{ lineHeight: `${height}px` }}
        >
          <span className={styles.percentage}>
            {percent}%
          </span>
          <span className={styles.value}>
            ({value})
          </span>
        </div>
      </div>
    );
  }
}

Bar.propTypes = {
  active: PropTypes.bool,
  limit: PropTypes.number.isRequired,
  value: PropTypes.number.isRequired,
  height: PropTypes.number,
};

Bar.defaultProps = {
  active: false,
  height: 32,
};

export default Bar;
