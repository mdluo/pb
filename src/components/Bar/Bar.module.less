.bar {
  margin-bottom: 20px;
  position: relative;
  outline: 0;
  border-radius: 4px;
  transition: all 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);

  &:global(.active) {
    box-shadow: 0 0 1px 2px rgba(47, 84, 235, 0.4);
  }

  &:global(.full) {
    :global(.ant-progress-bg) {
      border-top-right-radius: 4px;
      border-bottom-right-radius: 4px;
    }
  }

  :global {
    .ant-progress-bg,
    .ant-progress-inner {
      border-radius: 4px;
    }

    .ant-progress-bg {
      background-color: #adc6ff;
      border-top-right-radius: 0;
      border-bottom-right-radius: 0;
    }

    .ant-progress-status-exception .ant-progress-bg {
      background-color: #ffccc7;
    }
  }
}

.reading {
  width: 100%;
  position: absolute;
  text-align: center;
  top: 0;
}

.value {
  margin-left: 0.25em;
  font-size: 90%;
  color: #8c8c8c;
}
