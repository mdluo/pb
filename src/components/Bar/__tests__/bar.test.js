import React from 'react';
import { render, mount } from 'enzyme';

import { Progress } from 'antd';
import Bar from '..';

describe('Bar', () => {
  it('renders correctly', () => {
    const wrapper = render(<Bar value={50} limit={100} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders active correctly', () => {
    const wrapper = render(<Bar active value={50} limit={100} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('mounts one <Progress /> component', () => {
    const wrapper = mount(<Bar value={50} limit={100} />);
    expect(wrapper.find(Progress)).toHaveLength(1);
  });

  it('renders normal value (aliquot)', () => {
    const wrapper = render(<Bar value={50} limit={100} />);
    expect(wrapper.find('.percentage').text()).toBe('50%');
    expect(wrapper.find('.value').text()).toBe('(50)');
  });

  it('renders normal value (aliquant)', () => {
    const wrapper = render(<Bar value={50} limit={120} />);
    expect(wrapper.find('.percentage').text()).toBe('41%');
    expect(wrapper.find('.value').text()).toBe('(50)');
  });

  it('renders negative value', () => {
    const wrapper = render(<Bar value={-50} limit={100} />);
    expect(wrapper.find('.percentage').text()).toBe('0%');
    expect(wrapper.find('.value').text()).toBe('(0)');
  });

  it('throw on negative limit', () => {
    expect(() => {
      render(<Bar value={50} limit={-100} />);
    }).toThrow();
  });

  it('renders 100% value', () => {
    const wrapper = render(<Bar value={120} limit={120} />);
    expect(wrapper.find('.percentage').text()).toBe('100%');
    expect(wrapper.find('.value').text()).toBe('(120)');
    expect(wrapper.find('.ant-progress-status-exception')).toHaveLength(0);
  });

  it('renders over-limit value', () => {
    const wrapper = render(<Bar value={120} limit={100} />);
    expect(wrapper.find('.percentage').text()).toBe('120%');
    expect(wrapper.find('.value').text()).toBe('(120)');
    expect(wrapper.find('.ant-progress-status-exception')).toHaveLength(1);
  });
});
