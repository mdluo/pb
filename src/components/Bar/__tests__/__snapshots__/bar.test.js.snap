// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bar renders active correctly 1`] = `
<div
  class="bar active"
>
  <div
    class="ant-progress ant-progress-line ant-progress-status-normal ant-progress-default"
  >
    <div>
      <div
        class="ant-progress-outer"
      >
        <div
          class="ant-progress-inner"
        >
          <div
            class="ant-progress-bg"
            style="width:50%;height:32px"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="reading"
    style="line-height:32px"
  >
    <span
      class="percentage"
    >
      50%
    </span>
    <span
      class="value"
    >
      (50)
    </span>
  </div>
</div>
`;

exports[`Bar renders correctly 1`] = `
<div
  class="bar"
>
  <div
    class="ant-progress ant-progress-line ant-progress-status-normal ant-progress-default"
  >
    <div>
      <div
        class="ant-progress-outer"
      >
        <div
          class="ant-progress-inner"
        >
          <div
            class="ant-progress-bg"
            style="width:50%;height:32px"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="reading"
    style="line-height:32px"
  >
    <span
      class="percentage"
    >
      50%
    </span>
    <span
      class="value"
    >
      (50)
    </span>
  </div>
</div>
`;
